### ISSUE TITLE
"No code coverage driver is available" when running Infection with "variables_order" set to "EGPCS"

### ISSUE THREAD

**Issue by CViniciusSDias** [infection/infection#692]
| Question    | Answer
| ------------| ---------------
| Infection version | 0.12.2
| Test Framework version | PHPUnit 8.1.2
| PHP version | 7.3.2
| Platform    | Docker (image php:7.3-cli)
| Github Repo | - https://github.com/CViniciusSDias/google-crawler


I'm having trouble running infection inside my Docker container.

When I run it (`vendor/bin/infection`) I get the "Code Coverage does not exist" error even with xdebug enabled.

If I run the exact command it suggests (`'/app/vendor/phpunit/phpunit/phpunit' '--configuration' '/tmp/infection/phpunitConfiguration.initial.infection.xml'`) before and then specify the `--coverage` it works...

I tried adding `phpunit` to $PATH. Installed `infection` globally. Nothing worked.

PS: For a quicker test, remove (or comment) the "all" and "functional" testsuites and run it only with "unit" suite. It fails the same way but quicker.

<details>
 <summary>Output with issue</summary>
 
 ```
You are running Infection with Xdebug enabled.
     ____      ____          __  _
    /  _/___  / __/__  _____/ /_(_)___  ____
    / // __ \/ /_/ _ \/ ___/ __/ / __ \/ __ \
  _/ // / / / __/  __/ /__/ /_/ / /_/ / / / /
 /___/_/ /_/_/  \___/\___/\__/_/\____/_/ /_/

Running initial test suite...

PHPUnit version: 8.1.2

   28 [============================] < 1 sec
In CoverageDoesNotExistException.php line 55:
                                                                                                                                                                                         
  Code Coverage does not exist. File /tmp/infection/coverage-xml/index.xml is not found. Check phpunit version Infection was run with and generated config files inside /tmp/infection.  
  Command line: '/app/vendor/phpunit/phpunit/phpunit' '--configuration' '/tmp/infection/phpunitConfiguration.initial.infection.xml'                                                      
  Process Output: PHPUnit 8.1.2 by Sebastian Bergmann and contributors.                                                                                                                  
                                                                                                                                                                                         
  Random seed:   1557693865                                                                                                                                                              
  Error:         No code coverage driver is available                                                                                                                                    
                                                                                                                                                                                         
  .........................                                         25 / 25 (100%)                                                                                                       
                                                                                                                                                                                         
  Time: 29 ms, Memory: 4.00 MB                                                                                                                                                           
                                                                                                                                                                                         
  OK (25 tests, 28 assertions)                                                                                                                                                           
                                                                                                                                                                                         
   Make sure to either:                                                                                                                                                                  
  - Enable xdebug and run infection again                                                                                                                                                
  - Use phpdbg: phpdbg -qrr infection                                                                                                                                                    
  - Use --coverage option with path to the existing coverage report                                                                                                                      
  - Use --initial-tests-php-options option with `-d zend_extension=xdebug.so` and/or any extra php parameters
 ```
</details>

**Comment by bsacharski** [comment]
Question | Answer
-- | --
Infection version | 0.12.2
Test Framework version | PHPUnit 6.5.14
PHP version | 7.2.14
Platform | Mac OS (php installed via MAMP, configured with PATH)

I am having the exact problem. It would seem that unit tests are started without `xdebug` extension enabled:

<details>
  <summary>Infection output</summary>

```
You are running Infection with Xdebug enabled.
     ____      ____          __  _
    /  _/___  / __/__  _____/ /_(_)___  ____
    / // __ \/ /_/ _ \/ ___/ __/ / __ \/ __ \
  _/ // / / / __/  __/ /__/ /_/ / /_/ / / / /
 /___/_/ /_/_/  \___/\___/\__/_/\____/_/ /_/

Running initial test suite...

PHPUnit version: 6.5.14
    0 [>---------------------------] < 1 sec
    1 [->--------------------------] < 1 sec
    2 [-->-------------------------] < 1 sec
    3 [--->------------------------] < 1 sec
    4 [---->-----------------------] < 1 sec
    5 [----->----------------------]  1 sec
In CoverageDoesNotExistException.php line 55:
  Code Coverage does not exist. File /var/folders/n3/5rs817t110zct581csdz8mrc0000gn/T/infection/coverage-xml/index.xml is not found. Check phpunit version Infection was run with and generated config files inside /var/folders/n3/5rs817t110zct581csdz8mrc0000gn/T/infection.                                                                                                                             
  Command line: '/Users/beton/Development/shop/vendor/phpunit/phpunit/phpunit' '--configuration' '/var/folders/n3/5rs817t110zct581csdz8mrc0000gn/T/infection/phpunitConfiguration.initial.infection.xml'

  Process Output: PHPUnit 6.5.14 by Sebastian Bergmann and contributors.
  Error:         No code coverage driver is available
  ..                                                                  2 / 2 (100%)array (
    0 => 'Core',
    1 => 'date',
    2 => 'libxml',
    3 => 'openssl',
    4 => 'pcre',
    5 => 'sqlite3',
    6 => 'zlib',
    7 => 'bcmath',
    8 => 'bz2',
    9 => 'calendar',
    10 => 'ctype',
    11 => 'curl',
    12 => 'dom',
    13 => 'hash',
    14 => 'fileinfo',
    15 => 'filter',
    16 => 'ftp',
    17 => 'gd',
    18 => 'SPL',
    19 => 'iconv',
    20 => 'intl',
    21 => 'json',
    22 => 'ldap',
    23 => 'mbstring',
    24 => 'session',
    25 => 'standard',
    26 => 'mysqlnd',
    27 => 'PDO',
    28 => 'pdo_mysql',
    29 => 'pdo_sqlite',
    30 => 'Phar',
    31 => 'posix',
    32 => 'readline',
    33 => 'Reflection',
    34 => 'mysqli',
    35 => 'SimpleXML',
    36 => 'soap',
    37 => 'sockets',
    38 => 'sodium',
    39 => 'exif',
    40 => 'tokenizer',
    41 => 'wddx',
    42 => 'xml',
    43 => 'xmlreader',
    44 => 'xmlwriter',
    45 => 'xsl',
    46 => 'zip',
    47 => 'imap',
    48 => 'gettext',
    49 => 'pgsql',
    50 => 'pdo_pgsql',
    51 => 'Zend OPcache',
  )

  Time: 236 ms, Memory: 10.00MB

  OK (2 tests, 2 assertions)

   Make sure to either:
  - Enable xdebug and run infection again
  - Use phpdbg: phpdbg -qrr infection
  - Use --coverage option with path to the existing coverage report
  - Use --initial-tests-php-options option with `-d zend_extension=xdebug.so` and/or any extra php parameters                                                                                            
run [--test-framework TEST-FRAMEWORK] [--test-framework-options TEST-FRAMEWORK-OPTIONS] [-j|--threads THREADS] [--only-covered] [-s|--show-mutations] [--no-progress] [-c|--configuration CONFIGURATION] [--coverage COVERAGE] [--mutators MUTATORS] [--filter FILTER] [--formatter FORMATTER] [--min-msi MIN-MSI] [--min-covered-msi MIN-COVERED-MSI] [--log-verbosity LOG-VERBOSITY] [--initial-tests-php-options INITIAL-TESTS-PHP-OPTIONS] [--ignore-msi-with-no-mutations] [--debug]
```
</details>

However, when I start phpunit by itself, I can see that `xdebug` is loaded
<details>
  <summary>phpunit output</summary>

```

PHPUnit 6.5.14 by Sebastian Bergmann and contributors.

..                                                                                                                                                                                             2 / 2 (100%)array (
  0 => 'Core',
  1 => 'date',
  2 => 'libxml',
  3 => 'openssl',
  4 => 'pcre',
  5 => 'sqlite3',
  6 => 'zlib',
  7 => 'bcmath',
  8 => 'bz2',
  9 => 'calendar',
  10 => 'ctype',
  11 => 'curl',
  12 => 'dom',
  13 => 'hash',
  14 => 'fileinfo',
  15 => 'filter',
  16 => 'ftp',
  17 => 'gd',
  18 => 'SPL',
  19 => 'iconv',
  20 => 'intl',
  21 => 'json',
  22 => 'ldap',
  23 => 'mbstring',
  24 => 'session',
  25 => 'standard',
  26 => 'mysqlnd',
  27 => 'PDO',
  28 => 'pdo_mysql',
  29 => 'pdo_sqlite',
  30 => 'Phar',
  31 => 'posix',
  32 => 'readline',
  33 => 'Reflection',
  34 => 'mysqli',
  35 => 'SimpleXML',
  36 => 'soap',
  37 => 'sockets',
  38 => 'sodium',
  39 => 'exif',
  40 => 'tokenizer',
  41 => 'wddx',
  42 => 'xml',
  43 => 'xmlreader',
  44 => 'xmlwriter',
  45 => 'xsl',
  46 => 'zip',
  47 => 'imap',
  48 => 'gettext',
  49 => 'pgsql',
  50 => 'pdo_pgsql',
  51 => 'Zend OPcache',
  52 => 'xdebug',
)

Time: 638 ms, Memory: 6.00MB

OK (2 tests, 2 assertions)
```
</details>

**Comment by borNfreee** [comment]
Have you guys tried `master`? Does it fail with the same error?

**Comment by CViniciusSDias** [comment]
Yes! Same error with the infection/infection updated to `dev-master` version.

![output](https://uploaddeimagens.com.br/images/002/110/181/original/infection-error.png)

**Comment by CViniciusSDias** [comment]
> Have you guys tried `master`? Does it fail with the same error?

Yes! Same error with `infection/infection` dependency updated to `dev-master` version.

![output](https://uploaddeimagens.com.br/images/002/110/181/original/infection-error.png)

**Comment by sanmai** [comment]
What would `php vendor/bin/phpunit --coverage-text` tell?

**Comment by CViniciusSDias** [comment]
> What would `php vendor/bin/phpunit --coverage-text` tell?
> 
> If it can't make a coverage report, neither Infection can see it. Hence this will be not a problem of Infection.

Here's the output:

![phpunit output](https://uploaddeimagens.com.br/images/002/110/267/original/phpunit-coverage.png?1557893147)

PHPUnit CAN make the coverage report without any error. That's why I said in the issue description:
"If I run the exact command it suggests ('/app/vendor/phpunit/phpunit/phpunit' '--configuration' '/tmp/infection/phpunitConfiguration.initial.infection.xml') before and then specify the --coverage it works..."

=/

**Comment by sanmai** [comment]
Are you sure PHPUnit can write the coverage to disk?

Say, this should save the report:

    php vendor/bin/phpunit --coverage-html=coverage

**Comment by CViniciusSDias** [comment]
> Are you sure PHPUnit can write the coverage to disk?
> 
> Say, this should save the report:
> 
> ```
> php vendor/bin/phpunit --coverage-html=coverage
> ```

Yes. It can.

When I run the suggested command (`'/app/vendor/phpunit/phpunit/phpunit' '--configuration' '/tmp/infection/phpunitConfiguration.initial.infection.xml'`) the coverage is generated. Then if I send the path to the generated coverage to infection with the `--coverage` parameter it works.

`infection` isn't runing PHPUnit with xdebug enabled as @bsacharski said and showed.

**Comment by CViniciusSDias** [comment]
Doesn't seem to be adocker issue. @bsacharski mentioned he has the same issue and he's not using Docker.

**Comment by sanmai** [comment]
Neither I think this is an exclusive Docker issue, but since you can reliable reproduce it on Docker, that's where we going to stand for time being.

**Comment by sanmai** [comment]
Steps to reproduce the issue:
```
git clone https://github.com/CViniciusSDias/google-crawler
cd google-crawler/
rm tests/Functional/*
docker build -t php-google-crawler .
docker run -itv $(pwd):/app -w /app -u $(id -u):$(id -g) php-google-crawler composer install
docker run -itv $(pwd):/app -w /app -u $(id -u):$(id -g) php-google-crawler php vendor/bin/infection
```
(Functional tests are not required to reproduce the issue.)

**Comment by sanmai** [comment]
This issue boils down to `variables_order` having a non-standard setting (even though it is the default!). 

```
$ ./run php -r "var_dump(ini_get('variables_order'));"
Command line code:1:
string(5) "EGPCS"
```
If you remove this `E`, the issue will go away.

**Comment by CViniciusSDias** [comment]
If I add `-d variables_order=GPCS` to "initialTestsPhpOptions" in infection.json.dist it still doesn't work. Anyway, I'll wait for that PR to be merged and hope it solves it.

**Comment by sanmai** [comment]
This should work:
```
php -d variables_order=GPCS vendor/bin/infection
```

**Comment by CViniciusSDias** [comment]
It does! :-D

**Comment by borNfreee** [comment]
Thanks guys for creating this issue and reproducible repository.

@sanmai has fixed this bug, released as [0.13.1](https://github.com/infection/infection/releases/tag/0.13.1)
